### DISCUSSION TITLE
Python string -> nb?

### DISCUSSION THREAD

**Discussion by nschloe** [wjakob/nanobind#310] [Q&A] [answered]
Follow-up on https://github.com/wjakob/nanobind/discussions/309: How to pass Python strings to nb functions? The following Pybind11 function accepts a Python string:

```C++
void
foobar(const std::string & input_data) {
  // ...
}
```

but gives
```
TypeError: foobar(): incompatible function arguments. The following argument types are supported:
1. foobar(string: std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) -> bytes
```

in nb.

Any hints?

**Comment by wjakob** [comment, marked as answer]
You probably forgot to `#include` the STL type caster for strings. See the "Why another binding library" part of the docs for more detail on why those aren't included anymore.

**Comment by j-faria** [comment]
Sorry for hijacking this thread but I have a very related question.
I was trying to bind two overloaded C++ functions like

```cpp
void foobar(std::string s);
void foobar(std::vector<std::string> s);
```

and I noticed that nanobind always seems to call the second one, even when passed a single Python string. In that case, it builds a vector containing the individual characters. Is there a way to avoid this?

(I'm including both `<nanobind/stl/string.h>` and `<nanobind/stl/vector.h>` in case that matters)

**Reply by wjakob** [reply]
That's something I forgot about (pybind11 is more strict here). Fixed in https://github.com/wjakob/nanobind/commit/7e4a88b
